lib/vanagon/component.rb

```diff
@@ -190,16 +192,74 @@ def configfiles
       @files.select(&:configfile?)
     end
 
+    # @return [Set] a list of unique mirror URIs that should be used to
+    #   retrieve the upstream source before attempting to retrieve from
+    #   whatever URI was defined for #uri.
+    def mirrors
+      @mirrors ||= Set.new [deprecated_rewrite_url].compact
+    end
+
+    # Retrieve upstream source file from a mirror, by randomly iterating
+    # through #mirrors until there's no more mirrors left. Will #warn
+    # if the mirror's URI cannot be resolved or if the URI cannot
+    # be retrieved. Does not suppress any errors from
+    # Vanagon::Component::Source.
+    #
+    # @return [Boolean] return True if the source can be retrieved,
+    #   or False otherwise. This is because because each subclass of
+    #   Vanagon::Component::Source returns an inconsistent value
+    #   if #fetch is successful.
+    def fetch_mirrors(options)
+      mirrors.to_a.shuffle.each do |mirror|
```

Comment:
I'll try to explain it here instead of IRL, because it occurs to me that "nope" is too glib an answer to be useful. I'm basing my position here on my experiences with load balancing infrastructures. A round-robin approach (which shuffling an array basically is) is a reliable way to ensure that you're not accidentally pinning to one server while the rest of them rot and fall over. By shuffling the array ourselves, the hope is to mitigate the risk of us accidentally pinning to one URL that works because we uploaded stuff to the right place but none of our other URLs work because we goofed.